operator/duties/scheduler.go

```diff
@@ -0,0 +1,342 @@
+		indicesChg:          opts.IndicesChg,
+		blockPropagateDelay: blockPropagationDelay,
+
+		handlers: []dutyHandler{
+			NewAttesterHandler(),
+			NewProposerHandler(),
+			NewSyncCommitteeHandler(),
+		},
+
+		ticker:   make(chan phase0.Slot),
+		reorg:    make(chan ReorgEvent),
+		waitCond: sync.NewCond(&sync.Mutex{}),
+	}
+	if s.builderProposals {
+		s.handlers = append(s.handlers, NewValidatorRegistrationHandler())
+	}
+	return s
+}
+
+type ReorgEvent struct {
+	Slot     phase0.Slot
+	Previous bool
+	Current  bool
+}
+
+func (s *Scheduler) Start(ctx context.Context, logger *zap.Logger) error {
+	logger = logger.Named(logging.NameDutyScheduler)
+	logger.Info("duty scheduler started")
+
+	// Subscribe to head events.  This allows us to go early for attestations & sync committees if a block arrives, as well as
+	// re-request duties if there is a change in beacon block.
+	s.pool = pool.New().WithContext(ctx).WithCancelOnError()
+	subscriptionCtx, subscribed := context.WithCancel(ctx)
+	s.pool.Go(func(ctx context.Context) error {
+		defer subscribed()
+		err := s.beaconNode.Events(ctx, []string{"head"}, s.HandleHeadEvent(logger))
+		if err != nil {
+			return fmt.Errorf("failed to subscribe to head events: %w", err)
+		}
+		return nil
+	})
+
+	var reorgChs []chan<- ReorgEvent
+	var indicesChangeChs []chan<- bool
+	for _, handler := range s.handlers {
+		handler := handler
+		slotTicker := make(chan phase0.Slot)
+		s.slotTicker.Subscribe(slotTicker)
+
+		indicesChangeCh := make(chan bool)
+		indicesChangeChs = append(indicesChangeChs, indicesChangeCh)
+		reorgCh := make(chan ReorgEvent)
+		reorgChs = append(reorgChs, reorgCh)
+		handler.Setup(
+			handler.Name(),
+			logger,
+			s.beaconNode,
+			s.network,
+			s.validatorController,
+			s.ExecuteDuties,
+			slotTicker,
+			reorgCh,
+			indicesChangeCh,
+		)
+
+		s.pool.Go(func(ctx context.Context) error {
+			// Wait for the head event subscription to complete before starting the handler.
+			<-subscriptionCtx.Done()
+			handler.HandleDuties(ctx)
+			return nil
+		})
+	}
+
+	s.slotTicker.Subscribe(s.ticker)
+	go s.SlotTicker(ctx)
+
+	go fanOut(ctx, s.indicesChg, indicesChangeChs)
+	go fanOut(ctx, s.reorg, reorgChs)
+
+	return nil
+}
+
+func (s *Scheduler) Wait() error {
+	return s.pool.Wait()
+}
+
+// TODO: use a fan out package or Prysm's event.Feed, because sends in this implementation are blocking
+func fanOut[T any](ctx context.Context, in <-chan T, subscribers []chan<- T) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case item := <-in:
+			for _, s := range subscribers {
+				// Fan out the message to all subscribers.
+				select {
+				case s <- item:
+				case <-ctx.Done():
+				}
+			}
+		}
+	}
+}
+
+// SlotTicker handles the "head" events from the beacon node.
+func (s *Scheduler) SlotTicker(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case slot := <-s.ticker:
+			delay := s.network.SlotDurationSec() / 3 /* a third of the slot duration */
```

Comment:
maybe this should be a const?

Reply:
do you mean move the `3` to be a const?

Reply:
@y0sher @olegshmuelov 
this is called `INTERVALS_PER_SLOT` in the Beacon spec

https://github.com/prysmaticlabs/prysm/blob/2674e204cab117b1e902c78d72d8b6faaec7e6a0/config/params/mainnet_config.go#L119

Reply:
adjusted in https://github.com/bloxapp/ssv/pull/1044/commits/4c8d7399dc6536a36faba4c4e20aa4db22adf148